From de58a9554b0794ae549a3c495960512420650f66 Mon Sep 17 00:00:00 2001
From: Danesh M <daneshm90@gmail.com>
Date: Wed, 19 Aug 2015 15:43:48 -0700
Subject: [PATCH 37/74] FMRadio : Keep track of scanned frequencies in service

By keeping track of scanned frequencies in the service, this allows
us to fetch it when the activity is resumed and not rely on having
active callbacks.

Change-Id: Idea9cab3e5fd2a789333c3f126416e0102613aa8
---
 fmapp2/src/com/caf/fmradio/FMRadio.java       | 33 ++++++++++---------
 .../src/com/caf/fmradio/FMRadioService.java   | 15 +++++++++
 .../src/com/caf/fmradio/IFMRadioService.aidl  |  1 +
 3 files changed, 33 insertions(+), 16 deletions(-)

diff --git a/fmapp2/src/com/caf/fmradio/FMRadio.java b/fmapp2/src/com/caf/fmradio/FMRadio.java
index 18242ce..9865900 100644
--- a/fmapp2/src/com/caf/fmradio/FMRadio.java
+++ b/fmapp2/src/com/caf/fmradio/FMRadio.java
@@ -250,9 +250,6 @@ public class FMRadio extends Activity
    private ScrollerText mRadioTextScroller = null;
    private ScrollerText mERadioTextScroller = null;
 
-   /* Scanning frequencies */
-   ArrayList<Integer> mScannedFrequencies;
-
    private PresetStation mTunedStation = new PresetStation("", 102100);
    private PresetStation mPresetButtonStation = null;
 
@@ -530,7 +527,7 @@ public class FMRadio extends Activity
         }
         try {
             if (!mService.isSearchInProgress()) {
-                resetSearch();
+                mServiceCallbacks.onSearchComplete();
             }
         }catch (RemoteException e) {
             e.printStackTrace();
@@ -1780,9 +1777,6 @@ public class FMRadio extends Activity
       SharedPreferences.Editor editor = sp.edit();
       editor.clear();
       editor.commit();
-      if (mScannedFrequencies != null) {
-         mScannedFrequencies.clear();
-      }
    }
    public boolean fmConfigure() {
       boolean bStatus = true;
@@ -2141,13 +2135,19 @@ public class FMRadio extends Activity
    }
 
    private void saveStations() {
-       if (mScannedFrequencies != null && mScannedFrequencies.size() > 0) {
-           Collections.sort(mScannedFrequencies);
+       List<Integer> scannedFrequencies = null;
+       try {
+           scannedFrequencies = mService.getScannedFrequencies();
+       } catch (RemoteException e) {
+           e.printStackTrace();
+       }
+       if (scannedFrequencies != null && scannedFrequencies.size() > 0) {
+           Collections.sort(scannedFrequencies);
            SharedPreferences sp = getSharedPreferences(SCAN_STATION_PREFS_NAME, 0);
            SharedPreferences.Editor editor = sp.edit();
 
            int index = 0;
-           for (Integer freq : mScannedFrequencies) {
+           for (Integer freq : scannedFrequencies) {
                index++;
                editor.putString(STATION_NAME + index, index + "");
                editor.putInt(STATION_FREQUENCY + index, freq);
@@ -3182,11 +3182,6 @@ public class FMRadio extends Activity
          Log.d(LOGTAG, "mServiceCallbacks.onTuneStatusChanged: ");
          if (mIsScaning) {
              Log.d(LOGTAG, "isScanning....................");
-             if (mScannedFrequencies == null) {
-                 mScannedFrequencies = new ArrayList<Integer>();
-             }
-
-             mScannedFrequencies.add(FmSharedPreferences.getTunedFrequency());
          }
          cleanupTimeoutHandler();
          mHandler.post(mUpdateStationInfo);
@@ -3216,7 +3211,13 @@ public class FMRadio extends Activity
       public void onSearchComplete() {
          Log.d(LOGTAG, "mServiceCallbacks.onSearchComplete :");
          if (mIsScaning) {
-             if (mScannedFrequencies != null && mScannedFrequencies.size() > 0) {
+             List<Integer> scannedFrequencies = null;
+             try {
+                 scannedFrequencies = mService.getScannedFrequencies();
+             } catch (RemoteException e) {
+                 e.printStackTrace();
+             }
+             if (scannedFrequencies != null && !scannedFrequencies.isEmpty()) {
                  mShowStationList = true;
              } else {
                  mHandler.post(new Runnable() {
diff --git a/fmapp2/src/com/caf/fmradio/FMRadioService.java b/fmapp2/src/com/caf/fmradio/FMRadioService.java
index 0169f3d..2e55d98 100644
--- a/fmapp2/src/com/caf/fmradio/FMRadioService.java
+++ b/fmapp2/src/com/caf/fmradio/FMRadioService.java
@@ -166,6 +166,7 @@ public class FMRadioService extends Service
    private FMDeathRecipient mFMdr;
 
    //PhoneStateListener instances corresponding to each
+   private ArrayList<Integer> mScannedFrequencies = new ArrayList<Integer>();
 
    private FmRxRdsData mFMRxRDSData=null;
    // interval after which we stop the service when idle
@@ -2272,6 +2273,11 @@ public class FMRadioService extends Service
          return(mService.get().isSearchInProgress());
       }
 
+      public List<Integer> getScannedFrequencies()
+      {
+         return(mService.get().getScannedFrequencies());
+      }
+
       public int getExtenCountryCode()
       {
          return(mService.get().getExtenCountryCode());
@@ -2736,6 +2742,10 @@ public class FMRadioService extends Service
       return(bStatus);
    }
 
+   public List<Integer> getScannedFrequencies() {
+       return mScannedFrequencies;
+   }
+
    public boolean isSearchInProgress() {
       int state = mReceiver.getFMState();
       return state == qcom.fmradio.FmTransceiver.FMState_Srch_InProg;
@@ -3005,6 +3015,8 @@ public class FMRadioService extends Service
     */
    public boolean scan(int pty)
    {
+      // Clear previously scanned frequencies
+      mScannedFrequencies.clear();
       boolean bCommandSent=false;
       if (mReceiver != null)
       {
@@ -3540,6 +3552,9 @@ public class FMRadioService extends Service
             if(mReceiver != null) {
                clearStationInfo();
             }
+            if (isSearchInProgress()) {
+                mScannedFrequencies.add(frequency);
+            }
             if(mCallbacks != null)
             {
                mCallbacks.onTuneStatusChanged();
diff --git a/fmapp2/src/com/caf/fmradio/IFMRadioService.aidl b/fmapp2/src/com/caf/fmradio/IFMRadioService.aidl
index 3359542..d03c672 100644
--- a/fmapp2/src/com/caf/fmradio/IFMRadioService.aidl
+++ b/fmapp2/src/com/caf/fmradio/IFMRadioService.aidl
@@ -77,6 +77,7 @@ interface IFMRadioService
     boolean isRtPlusSupported();
     boolean isA2DPConnected();
     boolean isSearchInProgress();
+    List getScannedFrequencies();
     boolean getIntfDetLowTh();
     boolean getIntfDetHighTh();
     boolean getRxRepeatCount();
-- 
2.17.1

